SHA3.cs

```diff
@@ -0,0 +1,637 @@
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static byte[] HashData(byte[] key, byte[] source);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static byte[] HashData(ReadOnlySpan<byte> key, ReadOnlySpan<byte> source);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static int HashData(ReadOnlySpan<byte> key, ReadOnlySpan<byte> source, Span<byte> destination);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static bool TryHashData(
+        ReadOnlySpan<byte> key,
+        ReadOnlySpan<byte> source,
+        Span<byte> destination,
+        out int bytesWritten);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static byte[] HashData(byte[] key, Stream source);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static byte[] HashData(ReadOnlySpan<byte> key, Stream source);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static int HashData(ReadOnlySpan<byte> key, Stream source, Span<byte> destination);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static ValueTask<byte[]> HashDataAsync(byte[] key, Stream source, CancellationToken cancellationToken = default);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static ValueTask<byte[]> HashDataAsync(ReadOnlyMemory<byte> key, Stream source, CancellationToken cancellationToken = default);
+
+    [UnsupportedOSPlatform("ios")]
+    [UnsupportedOSPlatform("tvos")]
+    [UnsupportedOSPlatform("android")]
+    [UnsupportedOSPlatform("browser")]
+    [UnsupportedOSPlatform("osx")]
+    public static ValueTask<int> HashDataAsync(
+        ReadOnlyMemory<byte> key,
+        Stream source,
+        Memory<byte> destination,
+        CancellationToken cancellationToken = default);
+}
+
+public abstract partial class Shake : IDisposable {
```

Comment:
OpenSSL does not give us an API to stream XOFs back, to the best of my ability to understand. Certainly not for OpenSSL 1.1.1. CNG does (`BCRYPT_HASH_DONT_RESET_FLAG`).

Since OpenSSL doesn't let us stream back, the API shape does not reflect that.

I posted used existing API terms for SHAKE like append, "hash" etc, as that seems to be what other API surfaces are doing (CNG and OpenSSL). In fact for both of them, you use the exact same APIs. The only exception is that OpenSSL needs you to use `EVP_DigestFinalXOF` for the final computation.

Reply:
Another thing I considered, but abandoned, was "Can we make SHAKE work in `IncrementalHash`"? I largely concluded it would look weird. Either we would need a `XOFOutputSize` property, or something, or we would need a `byte[] GetHashAndReset(int)`, which would be weird because it would be a meaningless API for non-XOF hashes.

So they are separate.

Not to mention, I want one-shots, so trying to bolt it on to IncrementalHash doesn't give us an allocation-free XOF.

So the SHAKE APIs are basically `IncrementalHash` + one-shots.